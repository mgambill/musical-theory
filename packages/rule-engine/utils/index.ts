import type { Modifier } from '../types/modifiers'
export * from './modifiers'

// Function to apply extracted modifiers to a value
export function applyModifiers(
  modifiersMap: Map<string, Modifier>,
  value: any,
  modifiers: { key: string; args: any[] }[],
) {
  if (!modifiers || modifiers.length === 0) return value

  for (const { key, args } of modifiers ?? []) {
    const modifier = modifiersMap.get(key)
    if (modifier && modifier.canModify(value)) {
      value = modifier.value(value, ...args)
    } else {
      console.warn(`Modifier "${key}" not found or cannot modify value`, {
        value,
        modifier,
        args,
        can: modifier ? modifier.canModify(value) : false,
      })
    }
  }

  return value
}

// Extracts the value from an object and parses modifiers
export function getValueByPath<T>(
  obj: T,
  path: string,
  modifiersMap: Map<string, Modifier>,
): [any, { key: string; args: any[] }[]] {
  const modifierPattern = /\.\s*([a-zA-Z_$][\w$]*)\s*\((.*?)\)/g

  let value: unknown = obj
  const modifiers: { key: string; args: unknown[] }[] = []

  // Extract property path without modifiers
  const basePath = path.replace(modifierPattern, '')
  const properties = basePath.split('.')

  for (const part of properties) {
    if (value === undefined || value === null) break
    if (part === 'length') {
      const modifierKey = 'length'
      if (modifiersMap.has(modifierKey)) {
        modifiers.push({ key: modifierKey, args: [] })
      }

      break
    }
    value = value[part]
  }

  // Extract modifiers
  let match
  while ((match = modifierPattern.exec(path)) !== null) {
    const modifierKey = match[1] // e.g., "toUpperCase"
    const args = match[2] // e.g., "0,1"
      ? match[2].split(',').map((arg) => (isNaN(Number(arg)) ? arg.trim() : Number(arg.trim())))
      : []

    if (modifiersMap.has(modifierKey)) {
      modifiers.push({ key: modifierKey, args })
    }
  }

  return [value, modifiers]
}

export function toMap<T>(input?: Map<string, T> | Record<string, T>): Map<string, T> {
  if (input instanceof Map) {
    return input
  }

  const map = new Map<string, T>()
  if (input && typeof input === 'object') {
    Object.entries(input).forEach(([key, value]) => {
      map.set(key, value)
    })
  }
  return map
}
