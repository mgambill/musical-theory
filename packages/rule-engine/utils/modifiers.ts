import type { Modifier } from '../types'

export function isDate(value: unknown): value is Date {
  if (value instanceof Date) return !isNaN(value.getTime())

  if (typeof value !== 'string') return false

  const date = new Date(value)
  return !isNaN(date.getTime())
}

export function calculateAge(dateString: string | Date, currentDate?: Date): number {
  currentDate ??= new Date()
  const birthDate = new Date(dateString)
  let age = currentDate.getFullYear() - birthDate.getFullYear()
  const monthDifference = currentDate.getMonth() - birthDate.getMonth()

  if (
    monthDifference < 0 ||
    (monthDifference === 0 && currentDate.getDate() < birthDate.getDate())
  ) {
    age--
  }

  return age
}

// @ts-ignore
export const modifiers = new Map<string, Modifier>([
  [
    'length',
    {
      canModify: (value) => typeof value === 'string' || Array.isArray(value),
      value: (value) => value.length,
    },
  ],
  [
    'toLowerCase',
    {
      canModify: (value) => typeof value === 'string',
      value: (value) => value.toLowerCase(),
    },
  ],
  [
    'toUpperCase',
    {
      canModify: (value) => typeof value === 'string',
      value: (value) => value.toUpperCase(),
    },
  ],
  [
    'age',
    {
      canModify: (value: string | Date) => {
        if (value === undefined || value === null) return false
        return isDate(value)
      },
      value: (value: string | Date, currentDate: Date) => calculateAge(value, currentDate),
    },
  ] as unknown as Modifier<string | Date, number, [Date?]>,
])
