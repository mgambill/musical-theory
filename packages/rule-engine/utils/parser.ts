import type { Condition, ConditionOperator } from '../types/condition'

/**
 * Token types recognized by the parser.
 */
type Token =
  | { type: 'identifier'; value: string } // e.g. $.model.age or model.age
  | { type: 'operator'; value: string } // e.g. ==, !=, >, <, BETWEEN, IN, LIKE, etc.
  | { type: 'logical'; value: string } // e.g. AND, OR, NOT (used as logical connectors)
  | { type: 'literal'; value: string } // e.g. quoted strings, numbers, booleans, {CurrentDate}
  | { type: 'paren'; value: '(' | ')' } // For grouping expressions
  | { type: 'comma'; value: ',' } // For separating IN list items
  | { type: 'negate'; value: '!' } // For negating conditions

/**
 * Set of supported type-check function names.
 */
const typeCheckFunctions = new Set(['IsArray', 'IsBool', 'IsUndefined', 'IsDefined', 'IsInteger', 'IsNull', 'IsNumber', 'IsObject', 'IsString', 'IsEmpty', 'IsNotEmpty'])

/**
 * Tokenize the input string.
 * Updated regex includes comma as a token and explicitly treats BETWEEN, IN, and LIKE as operator tokens.
 */
function tokenize(input: string): Token[] {
  const tokens: Token[] = []
  // The regex now matches commas; keywords like BETWEEN/IN/LIKE will fall into the [^\s(),]+ bucket.
  const regex = /\s*(=>|==|!=|>=|<=|>|<|~|!|\(|\)|,|AND|OR|NOT|[^\s(),]+)/gi
  let match: RegExpExecArray | null

  while ((match = regex.exec(input)) !== null) {
    const tokenStr = match[0].trim()
    if (!tokenStr) continue

    if (tokenStr === '(' || tokenStr === ')') {
      tokens.push({ type: 'paren', value: tokenStr as '(' | ')' })
    } else if (tokenStr === ',') {
      tokens.push({ type: 'comma', value: tokenStr })
    } else if (['BETWEEN', 'IN', 'LIKE', 'STARTSWITH', 'ENDSWITH', 'ANY', 'ALL'].includes(tokenStr.toUpperCase())) {
      tokens.push({ type: 'operator', value: tokenStr.toUpperCase() })
    } else if (/^(AND|OR|NOT)$/i.test(tokenStr)) {
      tokens.push({ type: 'logical', value: tokenStr.toUpperCase() })
    } else if (/^(!)$/i.test(tokenStr)) {
      tokens.push({ type: 'negate', value: '!' })
    } else if (/^(==|!=|>=|<=|>|<|~)$/.test(tokenStr)) {
      tokens.push({ type: 'operator', value: tokenStr })
    } else if ((tokenStr.startsWith("'") && tokenStr.endsWith("'")) || (tokenStr.startsWith('"') && tokenStr.endsWith('"')) || (tokenStr.startsWith('`') && tokenStr.endsWith('`'))) {
      tokens.push({ type: 'literal', value: tokenStr })
    } else if (tokenStr.startsWith('{') && tokenStr.endsWith('}')) {
      tokens.push({ type: 'literal', value: tokenStr })
    } else if (!isNaN(Number(tokenStr))) {
      tokens.push({ type: 'literal', value: tokenStr })
    } else {
      // Otherwise, treat as an identifier.
      tokens.push({ type: 'identifier', value: tokenStr })
    }
  }
  return tokens
}

/**
 * Parse an expression combining terms with logical operators.
 */
function parseExpression(tokens: Token[]): [Condition, Token[]] {
  let [left, remainingTokens] = parseTerm(tokens)
  while (remainingTokens.length > 0 && remainingTokens[0].type === 'logical') {
    const logicalToken = remainingTokens.shift()!
    if (logicalToken.value === 'NOT') {
      const [term, newRemaining] = parseTerm(remainingTokens)

      left = { operator: 'Not', conditions: [term] } as Condition
      remainingTokens = newRemaining
    } else {
      const [right, newRemaining] = parseTerm(remainingTokens)
      left = {
        operator: logicalToken.value === 'AND' ? 'And' : 'Or',
        conditions: [left, right],
      } as Condition
      remainingTokens = newRemaining
    }
  }

  simplifyBooleanCondition(left)

  return [left, remainingTokens]
}

function negateCondition(term: Condition) {
  if (term.operator === 'IsEmpty') {
    term.operator = 'IsNotEmpty'
    return true
  } else if (term.operator === 'IsNull') {
    term.operator = 'IsNotNull'
    return true
  } else if (term.operator === 'Like') {
    term.operator = 'NotLike'
    return true
  } else if (term.operator === 'In') {
    term.operator = 'NotIn'
    return true
  }
  return false
}

function simplifyBooleanCondition(left: Condition) {
  if (left.operator === 'EqualTo' && left.value === false) {
    left.operator = 'IsFalse'
    left.value = undefined
  } else if (left.operator === 'EqualTo' && left.value === true) {
    left.operator = 'IsTrue'
    left.value = undefined
  }
}

/**
 * Parse a term: either a parenthesized expression, a type-check function, or a simple condition.
 */
function parseTerm(tokens: Token[]): [Condition, Token[]] {
  if (tokens.length === 0) {
    throw new Error('Unexpected end of input')
  }

  const token = tokens[0]

  // Handle NOT as a prefix operator.
  if (token.value === 'NOT') {
    tokens.shift() // consume the NOT token
    const [term, remaining] = parseTerm(tokens)

    return [{ operator: 'Not', conditions: [term] }, remaining]
  }

  if (token.type === 'negate') {
    tokens.shift() // consume the negate token
    const [term, remaining] = parseTerm(tokens)
    if (negateCondition(term)) return [term, remaining]
  }

  // Check for type-check function call.
  if (token.type === 'identifier' && typeCheckFunctions.has(token.value)) {
    tokens.shift() // consume the function name token
    return parseTypeCheckCondition(token, tokens)
  }

  // Handle parenthesized expressions.
  if (token.type === 'paren' && token.value === '(') {
    tokens.shift() // consume '('
    const [expr, remaining] = parseExpression(tokens)
    if (remaining.length === 0 || remaining[0].type !== 'paren' || remaining[0].value !== ')') {
      throw new Error('Expected closing parenthesis')
    }
    remaining.shift() // consume ')'
    return [expr, remaining]
  }

  // Fallback to parsing a simple condition.
  return parseSimpleCondition(tokens)
}

/**
 * Parse a type-check condition.
 * Expects a token already consumed for the function name (e.g., IsNull) and then:
 *   '(' <property> ')'
 * Returns a Condition with operator equal to the function name and the property (with "$." stripped if present).
 */
function parseTypeCheckCondition(firstToken: Token, tokens: Token[]): [Condition, Token[]] {
  const openParen = tokens.shift()
  if (!openParen || openParen.type !== 'paren' || openParen.value !== '(') {
    throw new Error(`Expected '(' after type check function ${firstToken.value}`)
  }
  const propToken = tokens.shift()
  if (!propToken || propToken.type !== 'identifier') {
    throw new Error(`Expected property identifier inside ${firstToken.value}()`)
  }
  const property = propToken.value.startsWith('$.') ? propToken.value.slice(2) : propToken.value
  const closeParen = tokens.shift()
  if (!closeParen || closeParen.type !== 'paren' || closeParen.value !== ')') {
    throw new Error(`Expected ')' after property in ${firstToken.value}()`)
  }
  return [{ operator: firstToken.value as ConditionOperator, property } as Condition, tokens]
}


/**
 * Parse a simple condition in the form: identifier operator value.
 * Supports special syntax for BETWEEN, IN, and LIKE.
 */
function parseSimpleCondition(tokens: Token[]): [Condition, Token[]] {
  // Expect an identifier (property)
  const identifierToken = tokens.shift()
  if (!identifierToken || identifierToken.type !== 'identifier' ) {
    throw new Error('Expected property identifier')
  }

  const property = identifierToken.value.startsWith('$.') ? identifierToken.value.slice(2) : identifierToken.value

  if (tokens.length === 0) {
    throw new Error('Expected operator after property')
  }

  const opToken = tokens.shift()!
  if (opToken.value === 'NOT') {
    tokens.splice(0, 0, identifierToken)
    const [expr, rem] = parseSimpleCondition(tokens)
    if (negateCondition(expr)) return [expr, rem]
  } else if (opToken.type !== 'operator') {
    throw new Error('Expected operator token')
  }
  const opValue = opToken.value.toUpperCase()

  // Handle quantifier operators ANY and ALL.
  if (opValue === 'ANY' || opValue === 'ALL') {
    if (tokens.length === 0 || tokens[0].type !== 'paren' || tokens[0].value !== '(') {
      throw new Error(`Expected '(' after quantifier operator ${opValue}`)
    }
    tokens.shift() // consume '('
    const [expr, rem] = parseExpression(tokens)
    if (rem.length === 0 || rem[0].type !== 'paren' || rem[0].value !== ')') {
      throw new Error(`Expected closing ')' for quantifier operator ${opValue} -218`)
    }
    rem.shift() // consume ')'
    // Return condition with quantifier operator.
    // Capitalize first letter: "Any" or "All"
    const quantOp = opValue.charAt(0) + opValue.slice(1).toLowerCase()
    return [{ operator: quantOp as ConditionOperator, property, conditions: [expr] } as Condition, rem]
  }

  // Special handling for BETWEEN, IN, and LIKE
  if (opValue === 'BETWEEN') {
    // Expect: <property> BETWEEN <lowerValue> AND <upperValue>
    const lowerToken = tokens.shift()
    if (!lowerToken || (lowerToken.type !== 'literal' && lowerToken.type !== 'identifier')) {
      throw new Error('Expected literal or identifier for lower bound in BETWEEN -232')
    }
    const lower = parseValue(lowerToken.value)
    const andToken = tokens.shift()
    if (!andToken || andToken.type !== 'logical' || andToken.value !== 'AND') {
      throw new Error('Expected AND in BETWEEN expression -237')
    }
    const upperToken = tokens.shift()
    if (!upperToken || (upperToken.type !== 'literal' && upperToken.type !== 'identifier')) {
      throw new Error('Expected literal or identifier for upper bound in BETWEEN -241')
    }
    const upper = parseValue(upperToken.value)
    return [{ operator: 'Between', property, value: [lower, upper] }, tokens]
  } else if (opValue === 'IN') {
    // Expect: <property> IN ( <value1>, <value2>, ... )
    if (tokens.length === 0 || tokens[0].type !== 'paren' || tokens[0].value !== '(') {
      throw new Error("Expected '(' after IN")
    }
    tokens.shift() // consume '('
    const inValues: Array<string | number | boolean | Date> = []
    while (tokens.length > 0) {
      // @ts-ignore
      if (tokens[0].type === 'paren' && tokens[0].value === ')') {
        tokens.shift() // consume ')'
        break
      }
      const valToken = tokens.shift()!
      if (valToken.type === 'comma') continue
      if (valToken.type !== 'literal' && valToken.type !== 'identifier') {
        throw new Error('Expected literal or identifier in IN list -261')
      }
      inValues.push(parseValue(valToken.value))
    }
    return [{ operator: 'In', property, value: inValues }, tokens]
  } else if (opValue === 'LIKE') {
    // Expect: <property> LIKE <pattern>
    const patternToken = tokens.shift()
    if (!patternToken || (patternToken.type !== 'literal' && patternToken.type !== 'identifier')) {
      throw new Error('Expected literal or identifier for LIKE pattern -270')
    }
    const pattern = parseValue(patternToken.value)
    return [{ operator: 'Like', property, value: pattern }, tokens]
  } else {
    // For other operators, simply parse the next token as a value.
    const valueToken = tokens.shift()
    if (!valueToken || (valueToken.type !== 'literal' && valueToken.type !== 'identifier')) {
      throw new Error('Expected literal or identifier for value -278')
    }
    const value = parseValue(valueToken.value)
    return [{ operator: mapOperator(opValue), property, value } as Condition, tokens]
  }
}

/**
 * Map a raw operator string to a ConditionOperator.
 */
function mapOperator(op: string): ConditionOperator {
  switch (op) {
    case '==':
      return 'EqualTo'
    case '!=':
      return 'NotEqualTo'
    case '>':
      return 'GreaterThan'
    case '<':
      return 'LessThan'
    case '>=':
      return 'GreaterThanOrEqualTo'
    case '<=':
      return 'LessThanOrEqualTo'
    case '~':
      return 'In' // legacy syntax for IN if needed
    case 'STARTSWITH':
      return 'StartsWith'
    case 'ENDSWITH':
      return 'EndsWith'
    default:
      throw new Error(`Unknown operator: ${op}`)
  }
}

/**
 * Parse a literal token into the appropriate type.
 */
function parseValue(tokenValue: string): string | number | boolean | Date {
  if ((tokenValue.startsWith("'") && tokenValue.endsWith("'")) || (tokenValue.startsWith('"') && tokenValue.endsWith('"')) || (tokenValue.startsWith('`') && tokenValue.endsWith('`'))) {
    return tokenValue.slice(1, -1)
  }
  if (!isNaN(Number(tokenValue))) {
    return Number(tokenValue)
  }
  if (tokenValue.toLowerCase() === 'true') return true
  if (tokenValue.toLowerCase() === 'false') return false
  return tokenValue
}

/**
 * Main parse function that converts an input string into a Condition.
 */
export function parse(input: string): Condition {
  const tokens = tokenize(input)
  const [condition, remaining] = parseExpression(tokens)
  if (remaining.length > 0) {
    throw new Error('Unexpected tokens remaining: ' + JSON.stringify(remaining))
  }
  return condition
}
