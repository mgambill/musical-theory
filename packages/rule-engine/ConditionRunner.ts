import type { Condition, ConditionType, Modifier, ConditionHandler, HandlerArgs, TokenDefinition, ConditionResultSet, ModifierMap, Rule } from './types'

import * as tokens from './tokens'

import { getValueByPath, applyModifiers, toMap } from './utils'
import { modifiers } from './utils/modifiers'

//export const condition = <const T, const W = never>(opt: ConditionType<T, W>) => opt as ConditionType<T>
export const condition = <const T, const W = never>(o: Partial<Condition> | ConditionType<T, W>) => o as Condition
export const rule = (options: Partial<Rule>) => Object.assign({ trigger: 'advance', condition: null, actions: [] }, options) as Rule

// The 'and' function: accepts an array of values (or nested logical expressions)
export function and(...args: Condition[]): Condition {
  return {
    operator: 'And',
    conditions: args,
  }
}

// The 'or' function: accepts an array of values (or nested logical expressions)
export function or(...args: Condition[]): Condition {
  return {
    operator: 'Or',
    conditions: args,
  }
}

// The 'or' function: accepts an array of values (or nested logical expressions)
export function not(...args: Condition[]): Condition {
  return {
    operator: 'Not',
    conditions: args,
  }
}

export function isRule(object: any): object is Rule {
  return 'trigger' in object && 'actions' in object
}

export function isCondition(object: any): object is Condition {
  return 'operator' in object
}

export class ConditionRunner {
  handlers: Map<string, ConditionHandler>
  modifiers: Map<string, Modifier>

  constructor(definition: ConditionRunnerDefinition) {
    if (!definition) console.warn('No definition provided for ConditionRunner. This will result in an empty runner.')
    this.handlers = definition?.handlers ?? new Map()
    this.modifiers = definition?.modifiers ?? new Map()
  }

  getValueByPath<T>(obj: T, path: string): [any, { key: string; args: any[] }[]] {
    return getValueByPath(obj, path, this.modifiers)
  }

  public filter<T, C extends Rule | Condition>(source: T[], ...condition: C[]): T[] {
    return source.filter((item) => condition.every((c) => this.test(item, c)))
  }

  public evaluate<T, C extends Rule | Condition>(source: T, ...condition: C[]): ConditionResultSet<C>[] {
    return condition.map((item) => ({ source: item, result: this.test(source, item) }))
  }

  isOption(value: any): boolean {
    return value && value.hasOwnProperty('value') && value.hasOwnProperty('label')
  }

  public test<T>(item: T, conditionOrRule: Condition | Rule | undefined | null, strict: boolean = false): boolean | null {
    if (conditionOrRule === undefined || conditionOrRule === null) throw new Error('unable filter on undefined condition or rule')

    if (isRule(conditionOrRule)) {
      if (conditionOrRule.condition === undefined || conditionOrRule.condition === null) return null
      return this.internalTest(item, conditionOrRule.condition as Condition, strict)
    }

    if (isCondition(conditionOrRule)) return this.internalTest(item, conditionOrRule, strict)

    console.assert(false, 'unable filter on invalid condition or rule')

    return null
  }

  internalTest<T>(item: T, condition: Condition, strict: boolean = false): boolean | null {
    // debugger
    const { operator } = condition
    // For property-based conditions, we need to evaluate the condition against each item.
    if (operator === 'And' || operator === 'Or' || operator === 'Not') {
      return operator === 'And' ? condition.conditions!.every((c) => this.test(item, c)) : operator === 'Or' ? condition.conditions!.some((c) => this.test(item, c)) : !this.test(item, condition)
    }

    const handler = this.handlers.get(operator)

    if (!handler) {
      throw new Error(`No handler for operator "${operator}"`)
    }

    if ('when' in condition && condition.when) {
      if (this.test(item, condition.when, strict) === false) {
        console.debug('When condition not met')
        return null
      }
    }

    if ('property' in condition && condition.property !== undefined) {

      // eslint-disable-next-line
      let [value, modifiers] = this.getValueByPath(item, condition.property)

      value = applyModifiers(this.modifiers, value, modifiers)

      // if the value is not an option assume we are comparing value of the option
      if (this.isOption(value) && this.isOption(condition.value) === false) {
        value = value.value
      }

      if (value === undefined) {
        if (operator === 'IsUndefined') return true
        if (operator === 'IsDefined') return false
        if (operator === 'IsNotNull') return true
        return null
        //  return strict === false
      }

      console.log('handler', handler, item, value, condition)

      return handler({
        value: 'value' in condition ? condition.value : null,
        property: value,
      } as HandlerArgs)
    }

    throw new Error('Invalid condition structure.')
  }
}

export class ConditionRunnerDefinition {
  handlers: Map<string, ConditionHandler>
  modifiers: Map<string, Modifier>
  debug: boolean

  constructor(_handlers?: TokenDefinition[], modifiers?: Map<string, Modifier> | Record<string, Modifier>, debug: boolean = false) {
    this.handlers = new Map()
    _handlers?.forEach(this.registerHandler.bind(this))

    this.modifiers = toMap(modifiers)

    this.debug = debug
  }

  // Allow users to add runtime modifiers
  public registerModifier(name: string, mod: Modifier): void {
    this.modifiers.set(name, mod)
  }

  public registerHandler({ name, handler }: TokenDefinition): void {
    const wrapper = (options: HandlerArgs): boolean => {
      const result = handler(options)
      if (this.debug) console.log(`${options.value} %c${name}`, 'color:limegreen', options.property, result)
      return result
    }
    this.handlers.set(name, wrapper)
  }
}

const def = new ConditionRunnerDefinition(Object.values(tokens), modifiers)

export const conditionRunnerDefinitions = def
