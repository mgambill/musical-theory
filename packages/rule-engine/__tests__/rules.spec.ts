import { test, expect, describe } from 'vitest'
import { ConditionRunner } from '../ConditionRunner'
import { condition, expressionParser, rule } from '../'
import { applyModifiers, getValueByPath, modifiers } from '../utils'
import { conditionRunnerDefinitions as definition } from '../ConditionRunner'

const runner = new ConditionRunner(definition)

const rules = [
  rule({
    label: 'age-check',
    trigger: 'milestone',
    condition: condition({
      key: 'age-check',
      property: 'dob.age()',
      operator: 'LessThanOrEqualTo',
      value: 18,
    }),
    actions: [
      {
        type: 'alert',
        message: 'You must be at least 18 years old to apply for naturalization. Please contact us for assistance.',
        exit: true,
      },
    ],
  }),
  rule({
    label: 'crime-check',
    trigger: 'milestone',
    condition: condition({
      key: 'crime-check',
      property: 'crime',
      operator: 'IsTrue',
    }),
    actions: [
      {
        type: 'alert',
        message: 'Criminal records are not eligible for this application.',
        exit: true,
      },
    ],
  }),
]

test('modifiers correct', () => {
  expect(runner.modifiers).toMatchSnapshot()
})

describe('rule 1', () => {
  const cond = condition({
    key: 'age-check',
    property: 'dob.age()',
    operator: 'LessThanOrEqualTo',
    value: 18,
    when: expressionParser('IsDefined(dob)'),
  })

  test('`when` constraint not met', () => {
    // Create a ConditionRunnerDefinition with sample handlers and modifiers
    const datasource = {}
    const result = runner.evaluate(datasource, cond)

    expect(result[0].result).toBe(null)
  })

  test('`when` constraint met; rule failed', () => {
    // Create a ConditionRunnerDefinition with sample handlers and modifiers
    const datasource = { dob: '2000-01-01' }
    const result = runner.evaluate(datasource, cond)

    expect(result[0].result).toBe(false)
  })

  test('`when` constraint met; rule passed', () => {
    // Create a ConditionRunnerDefinition with sample handlers and modifiers
    const datasource = { dob: '2022-01-01' }
    const result = runner.evaluate(datasource, cond)

    expect(result[0].result).toBe(true)
  })
})

describe('condition boolean', () => {
  const cond = condition({
    key: 'crime-check',
    property: 'crime',
    operator: 'IsTrue',
  })

  test('result is null, when input is null or undefined', () => {
    // Create a ConditionRunnerDefinition with sample handlers and modifiers
    const datasource = { crime: undefined }
    const result = runner.evaluate(datasource, cond!)

    expect(result[0].result).toBe(false)
  })

  test('`when` constraint met; rule passed', () => {
    // Create a ConditionRunnerDefinition with sample handlers and modifiers
    const datasource = { crime: false }
    const result = runner.evaluate(datasource, cond!)

    expect(result[0].result).toBe(false)
  })

  test('`when` constraint met; rule passed', () => {
    // Create a ConditionRunnerDefinition with sample handlers and modifiers
    const datasource = { crime: true }
    const result = runner.evaluate(datasource, cond!)

    expect(result[0].result).toBe(true)
  })
})

describe('undefined dataset objects', () => {
  const datasource = {
    gender: [],
    marital: [],
  }

  const result = runner.evaluate(datasource, ...rules.map((x) => x.condition!))

  test('result', () => {
    expect(result).toMatchInlineSnapshot(`
      [
        {
          "result": false,
          "source": {
            "key": "age-check",
            "operator": "LessThanOrEqualTo",
            "property": "dob.age()",
            "value": 18,
          },
        },
        {
          "result": false,
          "source": {
            "key": "crime-check",
            "operator": "IsTrue",
            "property": "crime",
          },
        },
      ]
    `)
  })

  test('age check', () => {
    expect(result[0].result).toBe(false)
  })

  test('crime check', () => {
    expect(result[1].result).toBe(false)
  })
})

describe('undefined dataset objects', () => {
  const datasource = {
    gender: [],
    marital: [],
  }

  const result = runner.evaluate(datasource, ...rules)

  test('result', () => {
    expect(result).toMatchInlineSnapshot(`
      [
        {
          "result": false,
          "source": {
            "actions": [
              {
                "exit": true,
                "message": "You must be at least 18 years old to apply for naturalization. Please contact us for assistance.",
                "type": "alert",
              },
            ],
            "condition": {
              "key": "age-check",
              "operator": "LessThanOrEqualTo",
              "property": "dob.age()",
              "value": 18,
            },
            "label": "age-check",
            "trigger": "milestone",
          },
        },
        {
          "result": false,
          "source": {
            "actions": [
              {
                "exit": true,
                "message": "Criminal records are not eligible for this application.",
                "type": "alert",
              },
            ],
            "condition": {
              "key": "crime-check",
              "operator": "IsTrue",
              "property": "crime",
            },
            "label": "crime-check",
            "trigger": "milestone",
          },
        },
      ]
    `)
  })

  test('age check', () => {
    expect(result[0].result).toBe(false)
  })

  test('crime check', () => {
    expect(result[1].result).toBe(false)
  })
})

describe('rulset handling', () => {
  const datasource = {
    gender: [],
    marital: [],
    crime: false,
    dob: '2000-03-05T05:00:00.000Z',
  }

  const result = runner.evaluate(datasource, ...rules)

  test('runner modifiers', () => {
    expect(runner.modifiers).toMatchInlineSnapshot(`
      Map {
        "length" => {
          "canModify": [Function],
          "value": [Function],
        },
        "toLowerCase" => {
          "canModify": [Function],
          "value": [Function],
        },
        "toUpperCase" => {
          "canModify": [Function],
          "value": [Function],
        },
        "age" => {
          "canModify": [Function],
          "value": [Function],
        },
      }
    `)
  })

  test('getValueByPath', () => {
    let [value, modifiers] = getValueByPath(datasource, 'dob.age()', runner.modifiers)
    expect(modifiers).toMatchInlineSnapshot(`
      [
        {
          "args": [],
          "key": "age",
        },
      ]
    `)

    expect(value).toMatch(datasource.dob)

    const age = applyModifiers(runner.modifiers, value, [{ key: 'age', args: [new Date('2025-03-05T05:00:00.000Z')] }])
    expect(age).toBe(25)
  })

  test('age check', () => {
    expect(result[0].result).toBe(false)
  })

  test('crime check', () => {
    expect(result[1].result).toBe(false)
  })

  test('filter', () => {
    const datasource2 = {
      gender: [],
      marital: [],
      crime: true,
      dob: '2020-03-05T05:00:00.000Z',
    }
    const result2 = runner.filter(Object.values(datasource2), ...rules)

    expect(result2).toMatchSnapshot()
  })
})
