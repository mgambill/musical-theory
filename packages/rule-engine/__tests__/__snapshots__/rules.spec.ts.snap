// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`modifiers correct 1`] = `
Map {
  "length" => {
    "canModify": [Function],
    "value": [Function],
  },
  "toLowerCase" => {
    "canModify": [Function],
    "value": [Function],
  },
  "toUpperCase" => {
    "canModify": [Function],
    "value": [Function],
  },
  "age" => {
    "canModify": [Function],
    "value": [Function],
  },
}
`;

exports[`rulset handling >  1`] = `
[
  "2000-03-05T05:00:00.000Z",
  [
    {
      "args": [],
      "key": "age",
    },
  ],
]
`;

exports[`rulset handling > filter 1`] = `[]`;

exports[`rulset handling > get 1`] = `
[
  "2000-03-05T05:00:00.000Z",
  [
    {
      "args": [],
      "key": "age",
    },
  ],
]
`;

exports[`rulset handling > getValueByPOath 1`] = `
[
  "2000-03-05T05:00:00.000Z",
  [
    {
      "args": [],
      "key": "age",
    },
  ],
]
`;

exports[`rulset handling > getValueByPath 1`] = `
[
  "2000-03-05T05:00:00.000Z",
  [
    {
      "args": [],
      "key": "age",
    },
  ],
]
`;
